### PR TITLE
Format: add an out_width function to output device

### PR DESCRIPTION
**Pull request by Octachron**
This PRs add a new function to the ouput device of  `Format`: `out_width`.
This new function can be used to compute the rendered width of substrings when printing on the lower-level device.

The aim of this addition is to improve the rendering for unicode strings by  informing the `Format` rendering engine that the width of   `𒆍𒀭𒊏𒆠` is  8, or maybe 4, but not 16 .

However, this only allows approximate computation of  unicode width (even before taking in account font ligatures and other text shaping-level rendering detail).

Indeed,  the formatting engine of `Format` decomposes substrings at the seams of format string constructors, and then computes the width of each substring in isolation. This is at best an approximation for unicode strings for which the `width` function is not an additive morphism.

For instance, the underlying formatting engine considers that
```ocaml    
        width "@[<h>ᄒ@,ᅡ@,ᆫ@,ᄀ@,ᅳ@,ᆯ@]"
```    
is equal to the sum
 ```ocaml   
        width "ᄒ" + width "ᅡ"+ width "ᆫ" + width "ᄀ" + width "ᅳ" + width "ᆯ" (* = 12 *)
```  
  when it should be equal to the width of the joined characters:
 ```ocaml   
        width "한글" (* = 4 *)
 ```

### CODE REVIEW

**Comment by dbuenzli** [comment]
How does that interact with the formatting instructions that have been specified as `@<n>` (which I use e.g. with 0 to output ANSI styling escapes, or sometimes to output certain Unicode characters, also it would be nice to have a proper function for this directive) ?

**Comment by Octachron** [comment]
Explicit width information (as provided with `@<n>`, `pp_print_substring_as` ,or `pp_print__as`) takes priority over the `out_width` function.

**Comment by dbuenzli** [comment]
Thanks. One also needs some guarantee that Format does not split strings given to it across UTF-8 encodings of characters which is not entirely clear to me.

Also of course what would be nice is a better default. A basic Unicode scalar value based `out_width` function like you have in the test[^1] already brings quite a bunch of Unicode characters into visual correctness. Any chance it could be installed by default ? 

[^1]: Which reminds me that I should probably [upstream this](https://github.com/dbuenzli/bytesrw/blob/3086187b08fc0e18d04abc4f9a81b24ec3c1d84b/test/utf8codec.ml#L11-L17) which is useful in many contexts.

> also it would be nice to have a proper function for this directive

Apparently I forgot it already existed since forever (?!). Not sure what made think it did not (perhaps I mixed with the fact that sometimes I wanted a form of @<%d> where the width would be taken as a parameter).

**Comment by Octachron** [comment]
At the very-least, we could have functions that install the ascii and "unicode scalar values" width for the various utf encodings.

Switching the default width has the small disadvantage that the byte length is a better width than the utf-8 scalar value one for texts encoded in UCS-2 (or utf-32)  (once the multiplicative factor corrected). However,  I am not sure if this situation happens, even more so since `js_of_ocaml` already assumes that on the OCaml side strings are utf-8 encoded.

**Comment by dbuenzli** [comment]
> At the very-least, we could have functions that install the ascii and "unicode scalar values" width for the various utf encodings.

That's not something I want to do in every of my programs…

I think you can safely forget about generating UTF-16 or UTF-32 with `Format`. Except in a few binary formats you'll never deal with that and for these cases generating UTF-8 and transcoding on the fly to the target encoding via `out_string` is a better option because… you'd need to be so careful – I suspect no use of any string literal/format whatsoever, no use of integer/float formatters, no use of any other library provided formatters etc. – that I'm pretty sure nobody does or would go into that.

Bottom line: Like it was eventually enshrined that `string` values if they are to represent text in OCaml should be UTF-8 encoded, enshrine that `Format` inputs UTF-8 encoded text. The system is consistent.

**Comment by Octachron** [comment]
I would be in favour of switching to an utf-8 unicode-aware width for `Format` by default. However, I am not sure if it would alleviate the need to install a better (or more specialized) width function in practice (to take in account combining characters, zero-width joiner, full-width characters, emoji, ...).

**Comment by dbuenzli** [comment]
We know anyways the problem is not solvable in general without interaction from the font rendering layer, so there's only so much you can do. But that would be a good gradual improvement. It will already do correctly on a bunch of cases like the [box drawing](https://www.unicode.org/charts/PDF/U2500.pdf) characters. Also lots of text is NFC in practice so you also get in quite a few alphabetic scripts.

I thought a bit about the backward compatibility problem. Assuming people used `Format` for what it, that is formatting text. I think the only people that would be negatively affected would be people who used `Format` to format `latin1` text (or any other single byte character encoding). Any other multibyte encoding would need explicit width instructions which means they would not be affected.

**Comment by Octachron** [comment]
Latin-1 and Latin-9 will only be affected at the margin: counting unicode scalar values  with replacement character only miscounts the number of latin-9 characters when the latin-9 stream contains an utf8 multibyte header followed by at least one continuation byte.

Assuming that a text is a stream of independent and identically distributed valid latin-9 bytes, a back-of-the-hand computation gives an error rate of -5%.  In practice, multibyte header and continuation byte seems anticorrelated and the error rate is much lower on the texts I have tested.

For instance, the project Gutenberg version of "Faust", "Pan Tedeusz", "Germinal" have no errors (the number of unicode scalar value with replacement is the same as the number of bytes) and I have seen an error rate of 0.011% (entirely due to "é»" and "à»") on the first tome of "Les Misérables".

**Comment by Octachron** [comment]
Let's thus install by default the "utf8 scalar values" width function (while keeping the original width function easily accessible).

File: stdlib/format.ml [hunk 1]
```diff
@@ -395,18 +399,19 @@ let format_pp_token state size = function
 
   | Pp_break { fits; breaks } ->
     let before, off, _ = breaks in
+    let before_width = pp_string_width state before in
```
Comment (gasche) [inline stdlib/format.ml]: If this is potentially slightly expensive, maybe move it in the `Some` case below where we need the result? (It does not sound completely bizarre to call, say, `Format.pp_print_text` at the toplevel.)

File: stdlib/format.ml [hunk 2]
```diff
@@ -953,6 +967,18 @@ let default_pp_mark_close_tag = function
 let default_pp_print_open_tag = ignore
 let default_pp_print_close_tag = ignore
 
+let utf8_scalar_width s ~pos ~len =
+  let rec width s count current last =
```
Comment (gasche) [inline stdlib/format.ml]: nitpick: to me `last` suggest that this position is included, while I read the code as excluding this position.

File: stdlib/format.mli [hunk 3]
```diff
@@ -918,6 +925,18 @@ val get_formatter_out_functions : unit -> formatter_out_functions
   @since 4.01
 *)
 
+val utf8_scalar_width: string -> pos:int -> len:int -> int
+(** [utf8_scalar_width s ~pos ~len] is the number of unicode scalar values in
+    the substring [String.sub s pos len]. Invalid byte sequences are implictly
+    replaced by [U+FFFD] since this yields a better width approximation for
+    other ascii-based encoding scheme like ISO-8859-15. This is the default
+    [out_width] function.
```
Comment (gasche) [inline stdlib/format.mli]: I would say "default [out_wdith] function since 5.4.", given that previous version did not have `out_width` but behaved as if they used `ascii_width`.

**Review by gasche — APPROVED** [review]
I cannot judge the design, but if @dbuenzli and @Octachron managed to agree then I feel confident that it is very reasonable. I looked at the implementation and it is fine, see minor comments. Approved.

**Comment by Octachron** [comment]
The PR (and https://github.com/ocaml/ocaml/pull/13569) is still missing a second standard library review. 
Maybe @nojb or @dra27 have an opinion on this matter?

File: stdlib/format.mli [hunk 4]
```diff
@@ -872,6 +873,13 @@ type formatter_out_functions = {
   It is called with a string [s], a start position [p], and a number of
   characters [n]; it is supposed to output characters [p] to [p + n - 1] of
   [s].
+- the [out_width] function informs the formatting engine of the width of the
+  substring as rendered on the output device. Explicit width information as
+  provided by [@<n>], {!pp_print_as} or {!pp_print_substring_as} takes priority
+  over this function. Moreover, the formatting engine evaluate the width of the
```
Comment (nojb) [inline stdlib/format.mli]:
```suggestion
  over this function. Moreover, the formatting engine evaluates the width of the
```

File: stdlib/format.mli [hunk 5]
```diff
@@ -886,6 +894,7 @@ type formatter_out_functions = {
   (e.g. {!Stdlib.output_string} and {!Stdlib.flush} for a
    {!Stdlib.out_channel} device, or [Buffer.add_substring] and
    {!Stdlib.ignore} for a [Buffer.t] output device),
+- fields [out_width] is the byte length of the substring
```
Comment (nojb) [inline stdlib/format.mli]:
```suggestion
- field [out_width] is the byte length of the substring
```

**Review by nojb — APPROVED** [review]
LGTM (modulo two typos)

Good to merge once the conflicts are resolved.

File: stdlib/format.mli (hunk 5 above)
Comment (Octachron) [reply stdlib/format.mli]: This was a bit more than a typo, since the documentation referred to the previous default [out_width] function. Thanks for noticing the typo !

**Comment by gasche** [comment]
Note: this PR, which I just merged, adds a function to compute unicode length in format.ml. @dbuenzli might want to look at potential factorization opportunities with #13760.

**Comment by dbuenzli** [comment]
If people want to I can follow up with adding `String.utf_8_uchar_length : ?pos:int -> ?len:int -> string -> int` or `String.utf_8_uchar_count`.

Note I think `Format.utf8_scalar_width` should be renamed to `Format.utf_8_scalar_width` for consistency with the rest of the stdlib.
